### DISCUSSION TITLE
Component Theming Feedback

### DISCUSSION THREAD

**Discussion by with-heart** [chakra-ui/chakra-ui#1707] [General]
Chakra v1 has added the ability to customize the theme via the component theming API, and to define custom components and composite components via `useStyleConfig` and `useMultiStyleConfig`. While we think these are powerful features, we also realize that there's a lot to them and they can be difficult to use or understand.

To help guide us to the best solutions for the community, we'd like for this issue to serve as a source of discussion. If you've felt pain around certain pieces of the API, or have relevant feedback or ideas you'd like to share, we'd love to hear it!

**Comment by PierreAndreis** [comment]
I think for me it was a bit confusing to understand at first, but once I got the jig of it I was able to really understand how powerful is and how it plays well when I just want to extend a theme

Example, this is my button component overriding on the theme
https://gist.github.com/PierreAndreis/e6fa613bf584040d9aca601129382695#file-button-ts-L8-L22

But, I'm still not sure if this is the right approach as this is not mentioned anywhere in the documentation.

**Comment by navin-moorthy** [comment]
For me, it was confusing a bit while starting, but once I understood it's customization power I love it.

But I have no idea on any better approach for this, I can see a similar kind of approach in [BumBag UI Kit](https://bumbag.style/theming/)

**Comment by Sebastiansc** [comment]
I had to dive in the code to understand how it works but once I did I really enjoy it. I like the `multiStyleConfig` specially. I think  something that could be improved would be allowing the parts to accept a function too, e.g.: 
`baseStyle: { item: props => ({...}) }` where `props` are the props provided directly to `MenuItem`. This way all style values  can live in the same place, otherwise we have to add  dynamic styles based  on local props within the component definition. This would be a pretty easy change, just editing `useMultiStyleConfig` to merge the `partsStyle[part]` as a function if any of the parts is a function

**Comment by ChiaJune** [comment]
Before diving into the source code, the documentation of Theme is very difficult to understand, especially for a novice like me who has never used css-in-js.  The other is, when I use `useStyleConfig` or `useMultiStyleConfig` to overwrite the component's style, it doesn't really overwrite the attribute with the same name, but adds the attribute with the same name.  I like chakra-ui, mainly because the components in it meet my needs. Thank you for your excellent work!

**Comment by with-heart** [comment]
Thanks for the feedback everyone! Sorry that we haven't responded to this, things have been hectic and short on time.

@ChiaJune have you had a chance to read through the new theming docs? @cereallarceny and @hericlesme rewrote them recently, and I'm wondering if the way they wrote them makes more sense to you than the old docs.

They start here: https://next.chakra-ui.com/docs/theming/customize-theme

**Reply by ChiaJune** [reply]
YES!
new theming docs is very clear.Thanks for great work!

**Reply by with-heart** [reply]
@ChiaJune Love to hear it!

**Comment by mdovn** [comment]
Hi, how can I overcome this limitation:
![image](https://user-images.githubusercontent.com/58581558/93076781-0f540e80-f6b2-11ea-822b-5fb71557e001.png)
I just need to modify the base theme a little. But currently, I have to write a lot of css once I use the function option.
btw, some component styles in the base theme also use the function for styling. They are not merged with my custom styles

**Reply by waptik** [reply]
> Hi, how can I overcome this limitation:
> ![image](https://user-images.githubusercontent.com/58581558/93076781-0f540e80-f6b2-11ea-822b-5fb71557e001.png)
> I just need to modify the base theme a little. But currently, I have to write a lot of css once I use the function option.
> btw, some component styles in the base theme also use the function for styling. They are not merged with my custom styles

How is your custom theme defined? and how did you try to use extendTheme?

**Comment by waptik** [comment]
I would like to say that the component theming feature is a really great one. Previously to achieve that, one would have to create a new component that extends the one the they want to modify, then write lots of bruhaha alongside some condition statements. But now it can be done by just adding the component name under component object in the custom theme and working with either or both of these object keys `baseStyle, sizes,variants,defaultProps,` under the component in question. And also, one shouldn't forget `useStyleConfig` and `useMultiStyleConfig` which helps in applying the customization.

I'd like to added that the `themeKey` (the key to use in `theme.components`) doesn't have to necessarily be named after any component in chakra. For example, i use `ButtonStyles` as themeKey(name) and reference it like this `useStyleConfig("ButtonStyles",, { size, variant })` and it worked.

**Comment by samageloff** [comment]
Yea, this is proving to be a really flexible and scalable solution. Really digging it so far. Pls don't change it, lol
